File: EPI/9.2/swift/9.2.swift
// swiftlint:disable conditional_binding_cascade

enum Token {
    case Operator((Int, Int) -> Int)
    case Number(Int)

    init?(string: String) {
        switch string {
        case "+":
            self = .Operator(+)
        case "-":
            self = .Operator(-)
        case "*":
            self = .Operator(*)
        case "/":
            self = .Operator(/)
        default:
            if let num = Int(string) {
                self = .Number(num)
            } else {
                return nil
            }
        }
    }
}

public func evaluate(expression: String) -> Int? {
    let tokens = expression.characters.split(",").map { Token(string: String($0))! }
    var stack = [Token]()

    for token in tokens {
        switch token {
        case let .Operator(op):
            let rhsToken = stack.removeLast()
            let lhsToken = stack.removeLast()
            if case let .Number(lhs) = lhsToken, case let .Number(rhs) = rhsToken {
                stack.append(.Number(op(lhs, rhs)))
            } else {
                return nil
            }
        case .Number:
            stack.append(token)
        }
    }

    if case let .Number(num) = stack.first! {
        return num
    } else {
        return nil
    }
}

File: EPI/9.2/typescript/9.2.ts
type OperatorFunction = (a: number, b: number) => number;

type Element = {type: 'value', value: number } | {type: 'operator', fn: OperatorFunction};

const OPERATOR_FUNCTIONS: {[op: string]: OperatorFunction} = {
    '+': (a, b) => a + b,
    '-': (a, b) => a - b,
    '*': (a, b) => a * b,
    '/': (a, b) => a / b,
};

function parseElement(s: string): Element {
    const fn = OPERATOR_FUNCTIONS[s];
    return fn ? {type: 'operator', fn} : {type: 'value', value: parseInt(s, 10)};
}

export default function exaluate(rpn: string): number {
    const elements = rpn.split(',').map(parseElement);
    const stack: number[] = [];

    for (const element of elements) {
        if (element.type === 'value') {
            stack.push(element.value);
        } else {
            if (stack.length < 2) {
                throw new Error('Invalid RPN expression ' + rpn);
            }
            const [b, a] = [stack.pop()!, stack.pop()!];
            stack.push(element.fn(a, b));
        }
    }

    if (stack.length !== 1) {
        throw new Error('Invalid RPN expression ' + rpn);
    }

    return stack[0];
}
